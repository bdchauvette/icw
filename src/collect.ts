export async function* collect<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): AsyncIterable<T[]> {
  let collectedResults = [];
  for await (let result of iterable) collectedResults.push(result);
  yield collectedResults;
}
