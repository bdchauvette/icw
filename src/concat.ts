import { flat } from "./flat";
import { toIterable } from "./__internal__/toIterable";
import { IterableLike } from "./IterableLike";

export async function* concat<T, U>(
  iterableLike: IterableLike<T>,
  ...values: U[]
): AsyncIterableIterator<T | U> {
  yield* toIterable(iterableLike);
  yield* flat(values, 1);
}
