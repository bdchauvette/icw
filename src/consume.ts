export const consume = async <T>(
  iterable: Iterable<T> | AsyncIterable<T>
): Promise<void> => {
  // @ts-ignore
  for await (let result of iterable) continue;
};
