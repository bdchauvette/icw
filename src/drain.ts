export async function drain<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): Promise<void> {
  // @ts-ignore
  for await (let result of iterable) continue;
}
