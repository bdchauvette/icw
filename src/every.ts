import { IterableLike } from "./IterableLike";
import { withIndex } from "./withIndex";

export async function every<T>(
  iterableLike: IterableLike<T>,
  predicate: (value: T, index?: number) => boolean | Promise<boolean>,
  thisArg?: any
): Promise<boolean> {
  let iterableDidPass = false;

  for await (let [value, index] of withIndex(iterableLike)) {
    iterableDidPass = Reflect.apply(predicate, thisArg, [value, index]);
    if (!(await iterableDidPass)) return false;
  }

  return iterableDidPass;
}
