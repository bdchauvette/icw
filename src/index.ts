export { ICW } from "./ICW";

export { from } from "./from";
export { of } from "./of";

export { drain } from "./drain";
export { filter } from "./filter";
export { map } from "./map";
export { reject } from "./reject";
export { scan } from "./scan";
export { withIndex } from "./withIndex";
