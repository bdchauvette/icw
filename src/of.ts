/**
 * Creates an AsyncIterator that yields each argument.
 */
export function of<T>(...values: T[]): AsyncIterable<T> {
  return {
    async *[Symbol.asyncIterator](): AsyncIterableIterator<T> {
      yield* values;
    }
  };
}
