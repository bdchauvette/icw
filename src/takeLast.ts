import { isNumber } from "./__internal__/isNumber";
import { toIterable } from "./__internal__/toIterable";
import { IterableLike } from "./IterableLike";

export async function* takeLast<T>(
  iterableLike: IterableLike<T>,
  numToTake: number
): AsyncIterableIterator<T> {
  if (!isNumber(numToTake) || numToTake < 0) {
    throw new RangeError("numToTake must be a non-negative number");
  }

  if (numToTake === 0) return;

  let buffer = [];

  for await (let value of toIterable(iterableLike)) {
    buffer.push(value);
    if (buffer.length > numToTake) buffer.shift();
  }

  yield* buffer;
}
