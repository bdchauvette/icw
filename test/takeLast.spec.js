import { takeLast } from "../src/takeLast";

import { of } from "../src/of";
import { ArrayLike } from "./helpers/ArrayLike";

test("rejects on non-IterableLike input", async () => {
  expect.assertions(2);
  try {
    await takeLast(null, 1).next();
  } catch (error) {
    expect(error).toBeInstanceOf(TypeError);
    expect(error.message).toMatchInlineSnapshot(
      `"Must provide an iterable, async iterable, Array-like value, or a Promise."`
    );
  }
});

test.each`
  description      | numToTake
  ${"null"}        | ${null}
  ${"undefined"}   | ${undefined}
  ${"non-numeric"} | ${"-1"}
  ${"< 0"}         | ${-1}
`("rejects when targetIndex is $description", async ({ numToTake }) => {
  expect.assertions(2);
  try {
    await takeLast(of("foo"), numToTake).next();
  } catch (error) {
    expect(error).toBeInstanceOf(RangeError);
    expect(error.message).toMatchSnapshot();
  }
});

test("returns same async iterator", () => {
  expect.assertions(1);
  expect(takeLast(of())).toReturnSameAsyncIterator();
});

test("returns a closeable iterator", async () => {
  expect.assertions(1);
  await expect(takeLast(of())).toBeCloseableAsyncIterator();
});

test("lazily consumes provided IterableLike input", async () => {
  expect.assertions(1);
  await expect(_ => takeLast(_, 1)).toLazilyConsumeWrappedAsyncIterable();
});

test.each`
  inputType          | iterableLike                          | numToTake | expectedValues
  ${"AsyncIterable"} | ${of("foo", "bar", "baz")}            | ${2}      | ${["bar", "baz"]}
  ${"Iterable"}      | ${["foo", "bar", "baz"]}              | ${2}      | ${["bar", "baz"]}
  ${"ArrayLike"}     | ${new ArrayLike("foo", "bar", "baz")} | ${2}      | ${["bar", "baz"]}
`(
  "yields only the last `numToTake` values from $inputType input",
  async ({ iterableLike, numToTake, expectedValues }) => {
    expect.assertions(expectedValues.length);
    for await (let value of takeLast(iterableLike, numToTake)) {
      expect(value).toStrictEqual(expectedValues.shift());
    }
  }
);

test.each`
  inputType          | iterableLike                          | expectedValues
  ${"AsyncIterable"} | ${of("foo", "bar", "baz")}            | ${["foo", "bar", "baz"]}
  ${"Iterable"}      | ${["foo", "bar", "baz"]}              | ${["foo", "bar", "baz"]}
  ${"ArrayLike"}     | ${new ArrayLike("foo", "bar", "baz")} | ${["foo", "bar", "baz"]}
  ${"Promise"}       | ${Promise.resolve("foo")}             | ${["foo"]}
`(
  "yields all values when `numToTake` is larger than the number of items in $inputType input",
  async ({ iterableLike, expectedValues }) => {
    expect.assertions(expectedValues.length);
    for await (let value of takeLast(iterableLike, Number.POSITIVE_INFINITY)) {
      expect(value).toStrictEqual(expectedValues.shift());
    }
  }
);

test.each`
  inputType          | iterableLike
  ${"AsyncIterable"} | ${of()}
  ${"Iterable"}      | ${[]}
  ${"ArrayLike"}     | ${new ArrayLike()}
`(
  "yields no values when $inputType input has no items",
  async ({ iterableLike }) => {
    expect.assertions(1);
    let iterator = takeLast(iterableLike, 1);
    let result = await iterator.next();
    expect(result.done).toBe(true);
  }
);

test.each`
  inputType          | iterableLike
  ${"AsyncIterable"} | ${of("foo", "bar", "baz")}
  ${"Iterable"}      | ${["foo", "bar", "baz"]}
  ${"ArrayLike"}     | ${new ArrayLike("foo", "bar", "baz")}
  ${"Promise"}       | ${Promise.resolve("foo")}
`(
  "yields no values from $inputType input when numToTake is 0",
  async ({ iterableLike }) => {
    expect.assertions(1);
    let iterator = takeLast(iterableLike, 0);
    let result = await iterator.next();
    expect(result.done).toBe(true);
  }
);
