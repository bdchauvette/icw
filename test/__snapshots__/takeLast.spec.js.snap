// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejects when targetIndex is "< 0" 1`] = `"numToTake must be a non-negative number"`;

exports[`rejects when targetIndex is "non-numeric" 1`] = `"numToTake must be a non-negative number"`;

exports[`rejects when targetIndex is "null" 1`] = `"numToTake must be a non-negative number"`;

exports[`rejects when targetIndex is "undefined" 1`] = `"numToTake must be a non-negative number"`;
